File: loader/src/xmodem.rs
use core::{mem, ptr::addr_eq};
use cortex_m::peripheral;
use misc::{
    flash,
    systick,
    image::{ImageHeader, IMAGE_MAGIC_APP, IMAGE_MAGIC_UPDATER, IMAGE_TYPE_APP, IMAGE_TYPE_UPDATER },
};
use stm32f4 as pac;

// XMODEM constants
pub const SOH: u8 = 0x01;
pub const EOT: u8 = 0x04;
pub const ACK: u8 = 0x06;
pub const NAK: u8 = 0x15;
pub const CAN: u8 = 0x18;
pub const X_C: u8 = 0x43;

// Timeout values in millis
const PACKET_TIMEOUT_MS: u32 = 5000; // 5 sec for each packet
const C_RETRY_INTERVAL_MS: u32 = 3000;
const MAX_RETRIES: u8 = 10;
const PACKET_SIZE: usize = 133;
const DATA_SIZE: usize = 128;

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum XmodemState {
    Idle,
    SendingInitialC,
    WaitingForData,
    ReceivingData,
    ProcessingPacket,
    Error,
    Complete,
}

#[derive(Debug)]
pub enum XmodemError {
    InvalidPacket,         // Invalid packet structure
    SequenceError,         // Sequence number mismatch
    CrcError,              // CRC checksum mismatch
    Cancelled,             // Transfer cancelled by sender
    Timeout,               // Timeout waiting for data
    FlashWriteError,       // Error writing to flash
    InvalidMagic,          // Invalid magic number in firmware
    OlderVersion,          // Firmware version is older than current
    TransferComplete,      // Transfer completed successfully
}

pub struct XmodemManager {
    state: XmodemState,
    target_addr: u32,
    current_addr: u32,
    expected_packet_num: u8,
    last_poll_time: u32,
    buffer: [u8; PACKET_SIZE],
    buffer_index: usize,
    expected_magic: u32,
    expected_img_type: u8,
    packet_count: u16,
    next_byte_to_send: Option<u8>,
    retries: u8,
    first_packet_processed: bool,
    current_sector_base: u32,
}

impl XmodemManager {
    pub fn new() -> Self {
        Self {
            state: XmodemState::Idle,
            target_addr: 0,
            current_addr: 0,
            expected_packet_num: 1,
            last_poll_time: 0,
            buffer: [0; PACKET_SIZE],
            buffer_index: 0,
            expected_magic: 0,
            expected_img_type: 0,
            packet_count: 0,
            next_byte_to_send: None,
            retries: 0,
            first_packet_processed: false,
            current_sector_base: 0,
        }
    }

    pub fn start(&mut self, addr: u32) {
        self.state = XmodemState::SendingInitialC;
        self.target_addr = addr;
        self.current_addr = addr;
        self.expected_packet_num = 1;
        self.buffer_index = 0;
        self.packet_count = 0;
        self.retries = 0;
        self.first_packet_processed = false;
        self.next_byte_to_send = Some(X_C); 
        self.last_poll_time = systick::get_tick_ms();

        if addr == crate::APP_ADDR {
            self.expected_magic = IMAGE_MAGIC_APP;
            self.expected_img_type = IMAGE_TYPE_APP;
        } else if addr == crate::UPDATER_ADDR {
            self.expected_magic = IMAGE_MAGIC_UPDATER;
            self.expected_img_type = IMAGE_TYPE_UPDATER;
        } else {
            self.state = XmodemState::Error;
            return;
        }

        self.current_sector_base = addr;
        // ersae first sector at target address
        let peripherals: stm32f4::Peripherals = unsafe {
            pac::Peripherals::steal()
        };
        match flash::erase_sector(&peripherals, addr) {
            0 => { self.state = XmodemState::Error;},
            _ => {
                // success
            }
        }
    }

    pub fn process_byte(&mut self, byte: u8) -> Result<bool, XmodemError> {
        let current_time: u32 = systick::get_tick_ms();

        match self.state {
            XmodemState::Idle => {
                Ok(false)
            },

            XmodemState::SendingInitialC => {
                // keep checkign for SOH to start firmware reception
                if byte == SOH {
                    self.buffer[0] = byte;
                    self.buffer_index = 1;
                    self.state = XmodemState::ReceivingData;
                    self.last_poll_time = current_time;
                    Ok(false)
                } else if byte == CAN {
                    self.state = XmodemState::Error;
                    Err(XmodemError::Cancelled)
                } else {
                    // 3 sec timeout for sending 'C' again
                    if current_time.wrapping_sub(self.last_poll_time) >= C_RETRY_INTERVAL_MS {
                        self.next_byte_to_send = Some(X_C);
                        self.last_poll_time = current_time;
                        self.retries += 1;

                        if self.retries >= MAX_RETRIES {
                            self.state = XmodemState:: Error;
                            return Err(XmodemError::Timeout);
                        }

                        Ok(true)
                    } else {
                        Ok(false)
                    }
                }
            },

            XmodemState::WaitingForData => {
                // timeout check first
                if current_time.wrapping_sub(self.last_poll_time) >= PACKET_TIMEOUT_MS {
                    self.state = XmodemState::Error;
                    return Err(XmodemError::Timeout);
                }

                match byte {
                    SOH => {
                        self.buffer[0] = byte;
                        self.buffer_index = 1;
                        self.state = XmodemState::ReceivingData;
                        self.last_poll_time = current_time;
                        Ok(false)
                    },
                    EOT => {
                        self.state = XmodemState::Complete;
                        self.next_byte_to_send = Some(ACK);
                        Err(XmodemError::TransferComplete)
                    },
                    CAN => {
                        self.state = XmodemState::Error;
                        Err(XmodemError::Cancelled)
                    },
                    _ => Ok(false)
                }
            },

            XmodemState::ReceivingData => {
                // timeout check first
                if current_time.wrapping_sub(self.last_poll_time) >= PACKET_TIMEOUT_MS {
                    self.state = XmodemState::Error;
                    return Err(XmodemError::Timeout);
                }
                
                // Add byte to buffer
                self.buffer[self.buffer_index] = byte;
                self.buffer_index += 1;
                
                // Check if we have a complete packet
                if self.buffer_index == PACKET_SIZE {
                    self.state = XmodemState::ProcessingPacket;
                    self.process_packet()
                } else {
                    Ok(false)
                }
            },

            XmodemState::ProcessingPacket => {
                // will do it in other function
                Ok(false)
            },
            
            XmodemState::Error | XmodemState::Complete => {
                // TODO:
                Ok(false)
            }
        }
    }

    fn process_packet(&mut self) -> Result<bool, XmodemError> {
        // dissassemply the packet
        let packet_num: u8 = self.buffer[1];
        let packet_num_complement: u8 = self.buffer[2];
        
        if packet_num.wrapping_add(packet_num_complement) != 0xFF {
            self.state = XmodemState::WaitingForData;
            self.buffer_index = 0;
            self.next_byte_to_send = Some(NAK);
            return Err(XmodemError::SequenceError);
        }
        
        if packet_num != self.expected_packet_num {
            self.state = XmodemState::WaitingForData;
            self.buffer_index = 0;
            self.next_byte_to_send = Some(NAK);
            return Err(XmodemError::SequenceError);
        }
        
        // CRC verification
        let received_crc: u16 = ((self.buffer[PACKET_SIZE-2] as u16) << 8) | (self.buffer[PACKET_SIZE-1] as u16);
        let calculated_crc: u16 = self.calculate_crc16(&self.buffer[3..3+DATA_SIZE]);
        
        if received_crc != calculated_crc {
            self.state = XmodemState::WaitingForData;
            self.buffer_index = 0;
            self.next_byte_to_send = Some(NAK);
            return Err(XmodemError::CrcError);
        }
        
        // For the very first packet we need to check magic and version first
        if packet_num == 1 && !self.first_packet_processed {
            let mut data_copy = [0u8; DATA_SIZE];
            data_copy.copy_from_slice(&self.buffer[3..3+DATA_SIZE]);

            let result = self.process_first_packet(&data_copy);
            if result.is_err() {
                return result;
            }
        } else {
            // check if we need to erase the next sector
            let next_addr = self.current_addr + DATA_SIZE as u32;
            let current_sector_end: u32 = self.current_sector_base + 0x20000;
            
            if next_addr > current_sector_end {
                let peripherals: stm32f4::Peripherals = unsafe { pac::Peripherals::steal() };
                let next_sector_base: u32 = self.current_sector_base + 0x20000;
                
                match flash::erase_sector(&peripherals, next_sector_base) {
                    0 => {
                        self.state = XmodemState::Error;
                        return Err(XmodemError::FlashWriteError);
                    },
                    _ => {
                        self.current_sector_base = next_sector_base;
                    }
                }
            }
            
            // write data to flash
            let mut data_copy = [0u8; DATA_SIZE];
            data_copy.copy_from_slice(&self.buffer[3..3+DATA_SIZE]);
            
            let peripherals: stm32f4::Peripherals = unsafe { pac::Peripherals::steal() };
            let result: u8 = flash::write(&peripherals, &data_copy, self.current_addr);
            
            if result != 0 {
                self.state = XmodemState::Error;
                return Err(XmodemError::FlashWriteError);
            }
            
            // Update current address
            self.current_addr += DATA_SIZE as u32;
        }
        
        // success
        self.state = XmodemState::WaitingForData;
        self.buffer_index = 0;
        self.expected_packet_num = self.expected_packet_num.wrapping_add(1);
        self.packet_count += 1;
        self.next_byte_to_send = Some(ACK);
        self.last_poll_time = systick::get_tick_ms();
        
        Ok(true) // Need to send ACK
    }

    fn process_first_packet(&mut self, data: &[u8]) -> Result<bool, XmodemError> {
        // Check if we have enough data for a header
        if data.len() < mem::size_of::<ImageHeader>() {
            self.state = XmodemState::Error;
            return Err(XmodemError::InvalidPacket);
        }
        
        let header: &ImageHeader = unsafe {
            let header_ptr: *const ImageHeader = data.as_ptr() as *const ImageHeader;
            &*header_ptr
        };
        
        // Check magic number
        if header.image_magic != self.expected_magic {
            self.state = XmodemState::Error;
            return Err(XmodemError::InvalidMagic);
        }
        
        // Check image type
        if header.image_type != self.expected_img_type {
            self.state = XmodemState::Error;
            return Err(XmodemError::InvalidMagic);
        }
        
        // Compare versions
        let current_header_addr: *const ImageHeader = self.target_addr as *const ImageHeader;
        let current_header: Option<&ImageHeader> = unsafe { 
            // Only read if the address is not all 0xFF (erased flash)
            if *(self.target_addr as *const u32) != 0xFFFFFFFF {
                Some(&*current_header_addr)
            } else {
                None
            }
        };
        
        if let Some(existing_header) = current_header {
            if existing_header.is_valid() && !header.is_newer_than(existing_header) {
                self.state = XmodemState::Error;
                return Err(XmodemError::OlderVersion);
            }
        }
        
        // Write the first packet data to flash
        let peripherals: stm32f4::Peripherals = unsafe { pac::Peripherals::steal() };
        let result: u8 = flash::write(&peripherals, data, self.current_addr);
        
        if result != 0 {
            self.state = XmodemState::Error;
            return Err(XmodemError::FlashWriteError);
        }
        
        // Update current address for next packet
        self.current_addr += DATA_SIZE as u32;
        self.first_packet_processed = true;
        
        Ok(true)
    }

    fn calculate_crc16(&self, data: &[u8]) -> u16 {
        let mut crc: u16 = 0;
        
        for &byte in data {
            crc ^= (byte as u16) << 8;
            for _ in 0..8 {
                if (crc & 0x8000) != 0 {
                    crc = (crc << 1) ^ 0x1021;
                } else {
                    crc <<= 1;
                }
            }
        }
        
        crc
    }

    pub fn should_send_byte(&mut self) -> bool {
        match self.state {
            XmodemState::SendingInitialC => {
                let current_time: u32 = systick::get_tick_ms();
                if current_time.wrapping_sub(self.last_poll_time) >= C_RETRY_INTERVAL_MS || self.next_byte_to_send.is_some() {
                    if self.next_byte_to_send.is_none() {
                        self.next_byte_to_send = Some(X_C);
                    }
                    self.last_poll_time = current_time;
                    self.retries += 1;
                    
                    if self.retries >= MAX_RETRIES {
                        self.state = XmodemState::Error;
                    }
                    
                    true
                } else {
                    false
                }
            },
            _ => self.next_byte_to_send.is_some()
        }
    }

    pub fn get_response(&mut self) -> Option<u8> {
        let response: Option<u8> = self.next_byte_to_send;
        self.next_byte_to_send = None;
        response
    }

    pub fn get_state(&self) -> XmodemState {
        self.state
    }

    pub fn get_packet_count(&self) -> u16 {
        self.packet_count
    }
}

File: common/src/xmodem.c
#include "xmodem.h"

// Timeout values in ms
#define PACKET_TIMEOUT 5000
#define C_RETRY_INTERVAL 3000
#define MAX_RETRIES 10

#define DATA_SIZE 128

static uint16_t calculate_crc16(const uint8_t* data, size_t len) {
    uint16_t crc = 0;
    
    for (size_t i = 0; i < len; i++) {
        crc ^= (uint16_t)data[i] << 8;
        for (uint8_t j = 0; j < 8; j++) {
            if (crc & 0x8000) {
                crc = (crc << 1) ^ 0x1021;
            } else {
                crc <<= 1;
            }
        }
    }
    
    return crc;
}

void xmodem_init(XmodemManager_t* manager, const XmodemConfig_t* config) {
    memset(manager, 0, sizeof(XmodemManager_t));
    manager->state = XMODEM_STATE_IDLE;
    manager->config = *config;
    manager->header_size = config->image_hdr_size;
}

void xmodem_start(XmodemManager_t* manager, uint32_t addr) {
    manager->state = XMODEM_STATE_SENDING_INITIAL_C;
    manager->target_addr = addr;
    manager->current_addr = addr;
    manager->expected_packet_num = 1;
    manager->buffer_index = 0;
    manager->packet_count = 0;
    manager->retries = 0;
    manager->first_packet_processed = 0;
    manager->next_byte_to_send = XMODEM_C;
    manager->last_poll_time = HAL_GetTick();
    manager->total_data_received = 0;
    manager->actual_firmware_size = 0;
    manager->received_eot = 0;
    manager->first_sector_erased = 0;
    manager->expected_total_packets = 0;
    manager->last_packet_useful_bytes = 0;
    
    // Set magic based on address
    if (addr == manager->config.app_addr) {
        manager->expected_magic = IMAGE_MAGIC_APP;
        manager->expected_img_type = IMAGE_TYPE_APP;
    } else if (addr == manager->config.updater_addr) {
        manager->expected_magic = IMAGE_MAGIC_UPDATER;
        manager->expected_img_type = IMAGE_TYPE_UPDATER;
    } else if (addr == manager->config.loader_addr) {
        manager->expected_magic = IMAGE_MAGIC_LOADER;
        manager->expected_img_type = IMAGE_TYPE_LOADER;
    } else {
        manager->state = XMODEM_STATE_ERROR;
        return;
    }
    
    manager->current_sector_base = addr;
}

static int process_first_packet(XmodemManager_t* manager, const uint8_t* data) {
    // First check if data is enough for header
    if (DATA_SIZE < sizeof(ImageHeader_t)) {
        return 0;
    }
    
    // Parse header
    ImageHeader_t* header = (ImageHeader_t*)data;
    
    // Check magic
    if (header->image_magic != manager->expected_magic) {
        return 0;
    }
    
    // Check image type
    if (header->image_type != manager->expected_img_type) {
        return 0;
    }
    
    // Store firmware size
    if (header->data_size > 0) {
        manager->actual_firmware_size = header->data_size + manager->header_size;
        
        // Calculate total packets
        uint32_t total_bytes = header->data_size + manager->header_size;
        uint32_t full_packets = total_bytes / DATA_SIZE;
        uint32_t remainder = total_bytes % DATA_SIZE;
        
        manager->expected_total_packets = full_packets + (remainder > 0 ? 1 : 0);
        
        if (remainder > 0) {
            manager->last_packet_useful_bytes = remainder;
        } else {
            manager->last_packet_useful_bytes = DATA_SIZE;
        }
    } else {
        // If no size info provided then use hardcoded value
        manager->actual_firmware_size = 0x40000; // 256 KB
    }
    
    // Check version
    ImageHeader_t current_header;
    if (manager->target_addr != 0xFFFFFFFF) {
        memcpy(&current_header, (void*)manager->target_addr, sizeof(ImageHeader_t));
        
        // Check if it's newer
        if (is_image_valid(&current_header) && !is_newer_version(header, &current_header)) {
            return 0;
        }
    }
    
    // Erase first sector if not done yet
    if (!manager->first_sector_erased) {
        if (!flash_erase_sector(manager->target_addr)) {
            return 0; // failed
        }
        manager->first_sector_erased = 1;
    }
    
    // Write first packet data to flash
    if (!flash_write(manager->current_addr, data, DATA_SIZE)) {
        return 0;
    }
    
    manager->total_data_received = DATA_SIZE;
    manager->current_addr += DATA_SIZE;
    manager->first_packet_processed = 1;
    
    return 1;
}

static size_t find_useful_bytes(XmodemManager_t* manager, const uint8_t* data, uint8_t packet_num) {
    if (manager->expected_total_packets > 0 && 
        packet_num == manager->expected_total_packets && 
        manager->last_packet_useful_bytes > 0) {
        return manager->last_packet_useful_bytes;
    }
    return DATA_SIZE;
}

XmodemError_t xmodem_process_byte(XmodemManager_t* manager, uint8_t byte) {
    uint32_t current_time = HAL_GetTick();
    
    switch (manager->state) {
        case XMODEM_STATE_IDLE:
            return XMODEM_ERROR_NONE;
            
        case XMODEM_STATE_SENDING_INITIAL_C:
            // Check for SOH
            if (byte == XMODEM_SOH) {
                manager->buffer[0] = byte;
                manager->buffer_index = 1;
                manager->state = XMODEM_STATE_RECEIVING_DATA;
                manager->last_poll_time = current_time;
                return XMODEM_ERROR_NONE;
            } else if (byte == XMODEM_CAN) {
                manager->state = XMODEM_STATE_ERROR;
                return XMODEM_ERROR_CANCELLED;
            } else {
                // Spam 'C'
                if (current_time - manager->last_poll_time >= C_RETRY_INTERVAL) {
                    manager->next_byte_to_send = XMODEM_C;
                    manager->last_poll_time = current_time;
                    manager->retries++;
                    
                    if (manager->retries >= MAX_RETRIES) {
                        manager->state = XMODEM_STATE_ERROR;
                        return XMODEM_ERROR_TIMEOUT;
                    }
                    
                    return XMODEM_ERROR_NONE;
                }
                return XMODEM_ERROR_NONE;
            }
            
        case XMODEM_STATE_WAITING_FOR_DATA:
            // Check for timeout
            if (current_time - manager->last_poll_time >= PACKET_TIMEOUT) {
                manager->state = XMODEM_STATE_ERROR;
                return XMODEM_ERROR_TIMEOUT;
            }
            
            // Process received byte
            if (byte == XMODEM_SOH) {
                manager->buffer[0] = byte;
                manager->buffer_index = 1;
                manager->state = XMODEM_STATE_RECEIVING_DATA;
                manager->last_poll_time = current_time;
                return XMODEM_ERROR_NONE;
            } else if (byte == XMODEM_EOT) {
                // End of transmission
                manager->received_eot = 1;
                manager->state = XMODEM_STATE_COMPLETE;
                manager->next_byte_to_send = XMODEM_ACK;
                
                // Verify size
                if (manager->actual_firmware_size > 0) {
                    if (manager->total_data_received < manager->actual_firmware_size) {
                        manager->state = XMODEM_STATE_ERROR;
                        return XMODEM_ERROR_INVALID_PACKET;
                    }
                    
                    // Adjust addr if received more than expected
                    if (manager->current_addr > manager->target_addr + manager->actual_firmware_size) {
                        manager->current_addr = manager->target_addr + manager->actual_firmware_size;
                    }
                }
                
                return XMODEM_ERROR_TRANSFER_COMPLETE;
            } else if (byte == XMODEM_CAN) {
                manager->state = XMODEM_STATE_ERROR;
                return XMODEM_ERROR_CANCELLED;
            }
            return XMODEM_ERROR_NONE;
            
        case XMODEM_STATE_RECEIVING_DATA:
            // Check for timeout
            if (current_time - manager->last_poll_time >= PACKET_TIMEOUT) {
                manager->state = XMODEM_STATE_ERROR;
                return XMODEM_ERROR_TIMEOUT;
            }
            
            // Add byte to buffer
            manager->buffer[manager->buffer_index++] = byte;
            
            // Check if we have a complete packet
            if (manager->buffer_index == 133) {
                manager->state = XMODEM_STATE_PROCESSING_PACKET;
                
                // Process the packet
                uint8_t packet_num = manager->buffer[1];
                uint8_t packet_num_comp = manager->buffer[2];
                
                // Check packet number integrity
                if ((packet_num + packet_num_comp) != 0xFF) {
                    manager->state = XMODEM_STATE_WAITING_FOR_DATA;
                    manager->buffer_index = 0;
                    manager->next_byte_to_send = XMODEM_NAK;
                    return XMODEM_ERROR_SEQUENCE_ERROR;
                }
                
                // Check packet sequence
                if (packet_num != manager->expected_packet_num) {
                    manager->state = XMODEM_STATE_WAITING_FOR_DATA;
                    manager->buffer_index = 0;
                    manager->next_byte_to_send = XMODEM_NAK;
                    return XMODEM_ERROR_SEQUENCE_ERROR;
                }
                
                // Verify CRC
                uint16_t received_crc = (manager->buffer[131] << 8) | manager->buffer[132];
                uint16_t calculated_crc = calculate_crc16(&manager->buffer[3], DATA_SIZE);
                
                if (received_crc != calculated_crc) {
                    manager->state = XMODEM_STATE_WAITING_FOR_DATA;
                    manager->buffer_index = 0;
                    manager->next_byte_to_send = XMODEM_NAK;
                    return XMODEM_ERROR_CRC_ERROR;
                }
                
                // Process data based on packet number
                if (packet_num == 1 && !manager->first_packet_processed) {
                    // First packet contains header
                    if (!process_first_packet(manager, &manager->buffer[3])) {
                        manager->state = XMODEM_STATE_ERROR;
                        return XMODEM_ERROR_INVALID_MAGIC;
                    }
                } else {
                    // Regular data packet
                    if (packet_num == 2 && !manager->first_packet_processed) {
                        manager->state = XMODEM_STATE_ERROR;
                        return XMODEM_ERROR_INVALID_PACKET;
                    }
                    
                    // Get useful data length
                    size_t useful_bytes = find_useful_bytes(manager, &manager->buffer[3], packet_num);
                    
                    if (useful_bytes == 0) {
                        // Skip useless
                        manager->state = XMODEM_STATE_WAITING_FOR_DATA;
                        manager->buffer_index = 0;
                        manager->expected_packet_num++;
                        manager->packet_count++;
                        manager->next_byte_to_send = XMODEM_ACK;
                        manager->last_poll_time = current_time;
                        return XMODEM_ERROR_NONE;
                    }
                    
                    // Track received data
                    manager->total_data_received += useful_bytes;
                    
                    if (manager->actual_firmware_size > 0 && 
                        manager->total_data_received > manager->actual_firmware_size) {
                        // Adjust useful bytes if we're receiving more than needed
                        uint32_t excess = manager->total_data_received - manager->actual_firmware_size;
                        if (excess < useful_bytes) {
                            useful_bytes -= excess;
                            manager->total_data_received = manager->actual_firmware_size;
                        } else {
                            useful_bytes = 0;
                        }
                    }
                    
                    // Check if we need to erase next sector
                    uint32_t next_addr = manager->current_addr + useful_bytes;
                    uint32_t current_sector_end = manager->current_sector_base + 0x20000;
                    
                    if (next_addr > current_sector_end && useful_bytes > 0) {
                        uint32_t next_sector_base = manager->current_sector_base + 0x20000;
                        
                        if (!flash_erase_sector(next_sector_base)) {
                            manager->state = XMODEM_STATE_ERROR;
                            return XMODEM_ERROR_FLASH_WRITE_ERROR;
                        }
                        
                        manager->current_sector_base = next_sector_base;
                    }
                    
                    // Write data to flash
                    if (useful_bytes > 0) {
                        if (!flash_write(manager->current_addr, &manager->buffer[3], useful_bytes)) {
                            manager->state = XMODEM_STATE_ERROR;
                            return XMODEM_ERROR_FLASH_WRITE_ERROR;
                        }
                        
                        manager->current_addr += useful_bytes;
                    }
                }
                
                // Move to next packet
                manager->state = XMODEM_STATE_WAITING_FOR_DATA;
                manager->buffer_index = 0;
                manager->expected_packet_num++;
                manager->packet_count++;
                manager->next_byte_to_send = XMODEM_ACK;
                manager->last_poll_time = current_time;
                return XMODEM_ERROR_NONE;
            }
            return XMODEM_ERROR_NONE;
            
        case XMODEM_STATE_ERROR:
        case XMODEM_STATE_COMPLETE:
        default:
            return XMODEM_ERROR_NONE;
    }
}

int xmodem_should_send_byte(XmodemManager_t* manager) {
    uint32_t current_time = HAL_GetTick();
    
    if (manager->state == XMODEM_STATE_SENDING_INITIAL_C) {
        if (current_time - manager->last_poll_time >= C_RETRY_INTERVAL || manager->next_byte_to_send != 0) {
            if (manager->next_byte_to_send == 0) {
                manager->next_byte_to_send = XMODEM_C;
            }
            manager->last_poll_time = current_time;
            manager->retries++;
            
            if (manager->retries >= MAX_RETRIES) {
                manager->state = XMODEM_STATE_ERROR;
            }
            
            return 1;
        }
        return 0;
    }
    
    return manager->next_byte_to_send != 0;
}

uint8_t xmodem_get_response(XmodemManager_t* manager) {
    uint8_t response = manager->next_byte_to_send;
    manager->next_byte_to_send = 0;
    return response;
}

XmodemState_t xmodem_get_state(XmodemManager_t* manager) {
    return manager->state;
}

uint16_t xmodem_get_packet_count(XmodemManager_t* manager) {
    return manager->packet_count;
}

void xmodem_cancel_transfer(XmodemManager_t* manager) {
    manager->state = XMODEM_STATE_ERROR;
    manager->next_byte_to_send = 0;
}